CDTDatastore/touchdb/TDReplicator.m

```diff
@@ -438,22 +435,38 @@ - (void)stop
 
 - (void)stopped
 {
-    CDTLogInfo(CDTREPLICATION_LOG_CONTEXT, @"%@ STOPPED", self);
-    CDTLogInfo(CDTREPLICATION_LOG_CONTEXT, @"Replication: %@ took %.3f sec; error=%@", self,
-            CFAbsoluteTimeGetCurrent() - _startTime, _error);
-    self.running = NO;
+    @synchronized(self) {
+        // only want stopped to run once
+        if (!_running) {
```

Comment:
I'm a bit concerned that `_running` has such a mix of synch (S) and unsynch (U) access in different places:

Method | `stop` | `stopped` | `goOnline` | `readyIfRetry` | `addToInbox` | `addRevsToInbox`
--- | --- | --- | --- | --- | --- | ---
TDPuller | U | **S** | U | - | - | - 
TDReplicator | **S** | **S** | U | U | U | U 

Let me know if you're comfortable with this or maybe we should do a walkthrough of this particular flag?

Reply:
Good analysis (everyone likes a table), `[TDPuller stop]` should be sync because it also depends on some boolean status flags which may cause race conditions. It seems awkward for a sub-class and a super-class to both have sync for the same methods, but sometimes the super-class method is invoked directly so it's unavoidable.

The `stopped` methods are called by `stop` or `updateActive`, or via the sub-class. Because of the `updateActive` case, they need their own synchronisation. I think `updateActive` itself doesn't need any synchronisation because its main job is just to fire `stopped` when the task count gets to zero.

 `goOnline` and `goOffline` are set through `TDReachability` which I'm pretty sure is only single threaded - there appears to be a run-loop updating the network reachability status.

`retryIfReady` doesn't manipulate any of the boolean status flags but can kick off `beginReplicating` to start the whole thing over. It gets called from `updateActive` but only in the `continuous` state and if there were failed revisions. From the CDT layer we don't support continuous replications (this field is hardcoded to `NO` at construction time), so I think we can safely ignore it.

`addToInbox` is driven from the `_changes` feed so I guess there is only one thread modifying the underlying `TDBatchers`'s `inBox` which is an `NSMutableArray`. (see https://developer.apple.com/library/content/documentation/Cocoa/Conceptual/Multithreading/ThreadSafetySummary/ThreadSafetySummary.html for details of which Cocoa classes are thread safe/unsafe).

`addRevsToInbox` only seems to get called from the `TDPusher` sub-class. Again it seems to only be called in a single-threaded fashion from `processInbox`.

Reply:
Oh I meant to add, in summary I think that `[TDPuller stop]` should be sync but I think we are OK with the others.

Reply:
Sounds good, thanks for digging through it.

Reply:
@tomblench Thanks for explaining and writing this up.